xinference/model/llm/llm_family.py

```diff
@@ -132,3 +153,125 @@ def cache_from_huggingface(
         )
 
     return cache_dir
+
+
+def _is_linux():
+    return platform.system() == "Linux"
+
+
+def _has_cuda_device():
+    cuda_visible_devices = os.environ.get("CUDA_VISIBLE_DEVICES")
+    if cuda_visible_devices:
+        return True
+    else:
+        from xorbits._mars.resource import cuda_count
+
+        return cuda_count() > 0
+
+
+def get_user_defined_llm_families():
+    with UD_LLM_FAMILIES_LOCK:
+        return UD_LLM_FAMILIES.copy()
+
+
+def match_llm(
+    model_name: str,
+    model_format: Optional[str] = None,
+    model_size_in_billions: Optional[int] = None,
+    quantization: Optional[str] = None,
+    is_local_deployment: bool = False,
+) -> Optional[Tuple[LLMFamilyV1, LLMSpecV1, str]]:
+    """
+    Find an LLM family, spec, and quantization that satisfy given criteria.
+    """
+    user_defined_llm_families = get_user_defined_llm_families()
+
+    for family in BUILTIN_LLM_FAMILIES + user_defined_llm_families:
+        if model_name != family.model_name:
+            continue
+        for spec in family.model_specs:
+            if (
+                model_format
+                and model_format != spec.model_format
+                or model_size_in_billions
+                and model_size_in_billions != spec.model_size_in_billions
+                or quantization
+                and quantization not in spec.quantizations
+            ):
+                continue
+            if quantization:
+                return family, spec, quantization
+            else:
+                # by default, choose the most coarse-grained quantization.
+                # TODO: too hacky.
+                quantizations = spec.quantizations
+                quantizations.sort()
+                for q in quantizations:
+                    if (
+                        is_local_deployment
+                        and not (_is_linux() and _has_cuda_device())
+                        and q == "4-bit"
+                    ):
+                        logger.warning(
+                            "Skipping %s for non-linux or non-cuda local deployment .",
+                            q,
+                        )
+                        continue
+                    return family, spec, q
+    return None
+
+
+def register_llm(llm_family: LLMFamilyV1, persist: bool):
+    from .utils import is_valid_model_name
+
+    if not is_valid_model_name(llm_family.model_name):
+        raise ValueError(
+            f"Invalid model name {llm_family.model_name}. The model name must start with a letter"
+            f" or a digit, and can only contain letters, digits, underscores, or dashes."
+        )
+
+    with UD_LLM_FAMILIES_LOCK:
+        for family in BUILTIN_LLM_FAMILIES + UD_LLM_FAMILIES:
+            if llm_family.model_name == family.model_name:
```

Comment:
Duplicated `model_id` won't harm since HF will not cache a same model twice.

Also, we are going to support caching a model with an URI in the near future, in which case the `model_id` may not exists.